engine/test_activation_shapes.py

```diff
@@ -0,0 +1,228 @@
+                # check if this is an active node and on the border
+                nodes[row][col].on_border = True
+                border_list.append(node)
+                if leftmost_node_pos is None or col < leftmost_node_pos[1]:
+                    leftmost_node = node
+                    leftmost_node_pos = (row, col)
+    border_nodes = border_list
+    leftmost_node = leftmost_node
+    leftmost_node_pos = leftmost_node_pos
+    return border_nodes,leftmost_node,leftmost_node_pos, nodes
+
+##Return bottom most node that is activated in the right column
+def return_right_position(pos, border_nodes):
+    candidate_nodes = [node for node in border_nodes if node.y == pos[1]+1]
+    if (candidate_nodes == []):
+        return None
+    else:
+        node = min(candidate_nodes,key=lambda node: node.x)
+        return (node.x,node.y)
+
+##Computes the lawnmover points for the grid using the activated and border nodes
+def get_all_lawnmower_waypoints_adjustable(nodes,lefmost_node_pos,border_nodes):
+        class WaypointPhase(Enum):
+            DOWN = 1
+            TERMINATE = 2
+        rows = nodes.shape[0]
+        phase = WaypointPhase.DOWN
+        curr_pos = lefmost_node_pos
+        waypoints = []
+        waypoints.append(curr_pos)
+        while (phase != WaypointPhase.TERMINATE):
+            if (phase == WaypointPhase.DOWN):
+                new_pos = (curr_pos[0]+1,curr_pos[1])
+                if curr_pos[0]+1 == rows and nodes[new_pos].activated:
+                    waypoints.append(new_pos)
+                    right_pos = return_right_position(new_pos,border_nodes)
+                    if right_pos is not None:
+                        waypoints.append(right_pos)
+                        curr_pos = right_pos
+                        phase = WaypointPhase.DOWN
+                    else:
+                        phase = WaypointPhase.TERMINATE
+                elif nodes[new_pos].activated:
+                    waypoints.append(new_pos)
+                    curr_pos = new_pos
+                else:
+                    right_pos = return_right_position(new_pos,border_nodes)
+                    if right_pos is not None:
+                        waypoints.append(right_pos)
+                        curr_pos = right_pos
+                        phase = WaypointPhase.DOWN
+                    else:
+                        phase = WaypointPhase.TERMINATE
+        return waypoints
+
+           
+
+
+
+def main():
+    rows = int(input("Num rows: "))##Good value 30
+    cols = int(input("Num cols: "))## Good value 30
+    activation_type = input("Activation type: ")
+    plt.figure()
+    plt.title("Activated Nodes")
+    plt.xlim(-1, rows+1)
+    plt.ylim(-1, cols+1)
+    plt.grid()
+    nodes = np.empty([rows, cols], dtype=object)
+    for i in range(rows):
+        for j in range(cols):
+            nodes[i,j] = Node(False,i,j)
+            
+    if activation_type == "rectangle":
```

Comment:
Should we be able to make straight lines here? crashes if i do
Inputs:
Num rows: 30
Num cols: 30
Activation type: rectangle
Rectangle row start: 5
Rectangle row end: 5
Rectangle col start: 9
Rectangle col end: 9

Reply:
Also, I tried to make a small rectangle and it highlighted a single border node instead. Is this supposed to happen? Was expecting 4 red dots (but it seems end row is exclusive?)

Inputs:
Num rows: 30
Num cols: 30
Activation type: rectangle
Rectangle row start: 5
Rectangle row end: 6
Rectangle col start: 7
Rectangle col end: 8

<img width="752" alt="Screen Shot 2022-09-14 at 1 11 18 AM" src="https://user-images.githubusercontent.com/60713782/190064828-7345439b-8a26-4d86-9061-725bd57908d4.png">

Reply:
I think It does but my plotting doesn't